transport/http/jsonrpc/server.go

```diff
@@ -0,0 +1,202 @@
+}
+
+// EndpointCodecMap maps the Request.Method to the proper EndpointCodec
+type EndpointCodecMap map[string]EndpointCodec
+
+// ServerOption sets an optional parameter for servers.
+type ServerOption func(*Server)
+
+// ServerBefore functions are executed on the HTTP request object before the
+// request is decoded.
+func ServerBefore(before ...httptransport.RequestFunc) ServerOption {
+	return func(s *Server) { s.before = before }
+}
+
+// ServerAfter functions are executed on the HTTP response writer after the
+// endpoint is invoked, but before anything is written to the client.
+func ServerAfter(after ...httptransport.ServerResponseFunc) ServerOption {
+	return func(s *Server) { s.after = after }
+}
+
+// ServerErrorLogger is used to log non-terminal errors. By default, no errors
+// are logged. This is intended as a diagnostic measure. Finer-grained control
+// of error handling, including logging in more detail, should be performed in a
+// custom ServerErrorEncoder or ServerFinalizer, both of which have access to
+// the context.
+func ServerErrorLogger(logger log.Logger) ServerOption {
+	return func(s *Server) { s.logger = logger }
+}
+
+// ServerFinalizer is executed at the end of every HTTP request.
+// By default, no finalizer is registered.
+func ServerFinalizer(f httptransport.ServerFinalizerFunc) ServerOption {
+	return func(s *Server) { s.finalizer = f }
+}
+
+// ServeHTTP implements http.Handler.
+func (s Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodPost {
+		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		io.WriteString(w, "405 must POST\n")
+		return
+	}
+	ctx := s.ctx
+
+	if s.finalizer != nil {
+		iw := &interceptingWriter{w, http.StatusOK}
+		defer func() { s.finalizer(ctx, iw.code, r) }()
+		w = iw
+	}
+
+	for _, f := range s.before {
+		ctx = f(ctx, r)
+	}
+
+	// Decode the body into an  object
+	var req Request
+	err := json.NewDecoder(r.Body).Decode(&req)
+	if err != nil {
+		s.logger.Log("err", err)
+		rpcErrorEncoder(ctx, err, w)
+		return
+	}
+
+	// Get the endpoint and codecs from the map using the method
+	// defined in the JSON  object
+	ecm := s.ecm[req.Method]
+
+	// TODO: Need to handle unregistered methods
+
+	// Decode the JSON  "params"
+	reqParams, err := ecm.Decode(ctx, req.Params)
+	if err != nil {
+		s.logger.Log("err", err)
+		rpcErrorEncoder(ctx, err, w)
+		return
+	}
+
+	// Call the Endpoint with the params
+	response, err := ecm.Endpoint(ctx, reqParams)
+	if err != nil {
+		s.logger.Log("err", err)
+		rpcErrorEncoder(ctx, err, w)
+		return
+	}
+
+	for _, f := range s.after {
+		ctx = f(ctx, w)
+	}
+
+	res := Response{
+		Error: Error{},
+	}
+
+	// Encode the response from the Endpoint
+	resParams, err := ecm.Encode(ctx, response)
+	if err != nil {
+		s.logger.Log("err", err)
+		rpcErrorEncoder(ctx, err, w)
+		return
+	}
+
+	res.Result = resParams
+
+	json.NewEncoder(w).Encode(res)
```

Comment:
Similar to the Decode, but here my request would be to be able to set options on the encoder: Example to pretty print the response, I'd want to `SetIndent`